aggregator/election_mgr.go

```diff
@@ -0,0 +1,545 @@
+		electionState:            scope.Gauge("election-state"),
+	}
+}
+
+type goalState struct {
+	id    int64
+	state ElectionState
+}
+
+type resignAction struct {
+	errCh chan error
+}
+
+type electionManager struct {
+	sync.RWMutex
+
+	nowFn               clock.NowFn
+	logger              xlog.Logger
+	reportInterval      time.Duration
+	campaignOpts        services.CampaignOptions
+	electionOpts        services.ElectionOptions
+	campaignRetrier     xretry.Retrier
+	changeRetrier       xretry.Retrier
+	changeVerifyTimeout time.Duration
+	electionKeyFmt      string
+	leaderService       services.LeaderService
+	leaderValue         string
+
+	state                  electionManagerState
+	doneCh                 chan struct{}
+	electionKey            string
+	electionStateWatchable xwatch.Watchable
+	goalStateLock          sync.Mutex
+	goalStateID            int64
+	goalStateWatchable     xwatch.Watchable
+	resignCh               chan resignAction
+	sleepFn                sleepFn
+	metrics                electionManagerMetrics
+}
+
+// NewElectionManager creates a new election manager.
+func NewElectionManager(opts ElectionManagerOptions) ElectionManager {
+	instrumentOpts := opts.InstrumentOptions()
+	campaignOpts := opts.CampaignOptions()
+	campaignRetrier := xretry.NewRetrier(opts.CampaignRetryOptions().SetForever(true))
+	changeRetrier := xretry.NewRetrier(opts.ChangeRetryOptions().SetForever(true))
+	electionStateWatchable := xwatch.NewWatchable()
+	electionStateWatchable.Update(FollowerState)
+	return &electionManager{
+		nowFn:                  opts.ClockOptions().NowFn(),
+		logger:                 instrumentOpts.Logger(),
+		reportInterval:         instrumentOpts.ReportInterval(),
+		campaignOpts:           campaignOpts,
+		electionOpts:           opts.ElectionOptions(),
+		campaignRetrier:        campaignRetrier,
+		changeRetrier:          changeRetrier,
+		changeVerifyTimeout:    opts.ChangeVerifyTimeout(),
+		electionKeyFmt:         opts.ElectionKeyFmt(),
+		leaderService:          opts.LeaderService(),
+		leaderValue:            campaignOpts.LeaderValue(),
+		state:                  electionManagerNotOpen,
+		doneCh:                 make(chan struct{}),
+		electionStateWatchable: electionStateWatchable,
+		goalStateWatchable:     xwatch.NewWatchable(),
+		resignCh:               make(chan resignAction, 1),
+		sleepFn:                time.Sleep,
+		metrics:                newElectionManagerMetrics(instrumentOpts.MetricsScope()),
+	}
+}
+
+func (mgr *electionManager) Open(shardSetID string) error {
+	mgr.Lock()
+	defer mgr.Unlock()
+
+	if mgr.state != electionManagerNotOpen {
+		return errElectionManagerAlreadyOpenOrClosed
+	}
+	mgr.state = electionManagerOpen
+	mgr.electionKey = fmt.Sprintf(mgr.electionKeyFmt, shardSetID)
+
+	_, stateChangeWatch, err := mgr.goalStateWatchable.Watch()
+	if err != nil {
+		return err
+	}
+	go mgr.watchGoalStateChanges(stateChangeWatch)
+
+	_, verifyWatch, err := mgr.goalStateWatchable.Watch()
```

Comment:
Sure can do, this is called on aggregator startup, so if this fails, the process will exit, but doesn't hurt to be defensive.